From 1d7372ace8a12f132a99cd156af5a3d77b25727f Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Mon, 11 Nov 2019 11:19:33 +0200
Subject: [PATCH] [SW-6739] arcboot: remove unused symbols from elf

This is first measure to decrease binary size
We need to be at ROM/ICCM/FW_SRAM either are 64KB
Now size is 59KB (see size at BINARIES_DIR)
Change-Id: I0fccb36cac8d2fadf93881db98f7244b6a4b28c0
---

diff --git a/Makefile b/Makefile
index 94064a6..fdbdcef 100644
--- a/Makefile
+++ b/Makefile
@@ -10,16 +10,20 @@
 OBJCOPY = $(CROSS_COMPILE)objcopy
 
 LIBDIR = sdk/lib
-CFLAGS = -DSB_RSA_STD -DSB_FAST_MODE -DELP_SOURCE -Wall -W -O3
-CFLAGS += -Isdk/src
-CFLAGS += -Isdk/src/include
-CFLAGS += -Isdk/elpsoft/src/headers
-CFLAGS += -Isdk/elpsoft/src/math
+ARCBOOT_CFLAGS = -DSB_RSA_STD -DSB_FAST_MODE -DELP_SOURCE -Wall -W -O3
+ARCBOOT_CFLAGS += -Isdk/src
+ARCBOOT_CFLAGS += -Isdk/src/include
+ARCBOOT_CFLAGS += -Isdk/elpsoft/src/headers
+ARCBOOT_CFLAGS += -Isdk/elpsoft/src/math
 
 LDFLAGS = -flto
 LDFLAGS += -L $(LIBDIR) -lsb -lellipsys
 LDFLAGS += -T arcboot.lds
 LDFLAGS += -nostartfiles
+
+# This combination allows removing unused symbols from libsb.a and libellipsys.a
+LDFLAGS += -Wl,--gc-sections
+export CFLAGS = -Os -fdata-sections -ffunction-sections
 
 .PHONY: all libsb libellipsys sbtool
 all: libsb libellipsys
@@ -38,7 +42,7 @@
 	@$(CC) $^ $(LDFLAGS) -o $@
 
 %.o: %.c
-	@$(CC) $(CFLAGS) -c $< -o $@
+	@$(CC) $(ARCBOOT_CFLAGS) -c $< -o $@
 
 %.o: %.s
 	@$(CC) $(AFLAGS) -c $< -o $@
